File: src/ec2/disk.rs
use std::{
    fs::{self, File},
    io::{self, Write},
    thread, time,
};

/// Makes a new file system on a specified device.
///
/// e.g.,
/// sudo mkfs -t ext4 /dev/nvme1n1
///
/// ref. See https://github.com/cholcombe973/block-utils/blob/master/src/lib.rs for other commands.
/// ref. https://stackoverflow.com/questions/45167717/mounting-a-nvme-disk-on-aws-ec2
pub fn make_filesystem(filesystem_name: &str, device_name: &str) -> io::Result<(String, String)> {
    match run_make_filesystem(filesystem_name, device_name) {
        Ok(v) => {
            log::debug!("first time success 'make_filesystem'");
            return Ok(v);
        }
        Err(e) => {
            log::warn!(
                "first time failure 'make_filesystem' {}; retrying after 5-second...",
                e
            );
            thread::sleep(time::Duration::from_secs(5));
            return run_make_filesystem(filesystem_name, device_name);
        }
    }
}

pub fn run_make_filesystem(
    filesystem_name: &str,
    device_name: &str,
) -> io::Result<(String, String)> {
    let device_path = if device_name.starts_with("/dev/") {
        device_name.to_string()
    } else {
        format!("/dev/{}", device_name)
    };

    log::info!(
        "making file system with 'mkfs' command on the device path '{}'",
        device_path
    );

    let cmd = format!("sudo mkfs -t {} {}", filesystem_name, device_path);
    let res = command_manager::run(&cmd);
    if res.is_err() {
        // e.g., mke2fs 1.45.5 (07-Jan-2020) /dev/nvme1n1 is mounted; will not make a filesystem here!
        // TODO: handle "mount: /data: wrong fs type, bad option, bad superblock on /dev/nvme1n1, missing codepage or helper program, or other error."
        let e = res.err().unwrap();
        if !e
            .to_string()
            .contains(format!("{} is mounted", device_path).as_str())
        {
            return Err(e);
        }

        log::warn!("ignoring the 'is mounted' error '{}'", e.to_string());
        Ok((String::new(), e.to_string()))
    } else {
        res
    }
}

/// Mounts the file system to the specified directory.
/// And updates "/etc/fstab" to auto remount in case of instance reboot.
///
/// e.g.,
/// sudo mount /dev/nvme1n1 /data -t ext4
///
/// ref. See https://github.com/cholcombe973/block-utils/blob/master/src/lib.rs for other commands.
/// ref. https://stackoverflow.com/questions/45167717/mounting-a-nvme-disk-on-aws-ec2
pub fn mount_filesystem(
    filesystem_name: &str,
    device_name: &str,
    dir_name: &str,
) -> io::Result<(String, String)> {
    let device_path = if device_name.starts_with("/dev/") {
        device_name.to_string()
    } else {
        format!("/dev/{}", device_name)
    };

    log::info!(
        "mounting the file system with 'mount' command on the device path '{}'",
        device_path
    );

    let cmd = format!(
        "sudo mount {} {} -t {}",
        device_path, dir_name, filesystem_name
    );
    let res = command_manager::run(&cmd);
    if res.is_err() {
        // e.g., mount: /data: /dev/nvme1n1 already mounted on /data
        let e = res.err().unwrap();
        if !e
            .to_string()
            .contains(format!("{} already mounted", device_path).as_str())
        {
            return Err(e);
        }

        log::warn!("ignoring the 'already mounted' error '{}'", e.to_string());
        Ok((String::new(), e.to_string()))
    } else {
        res
    }
}

const FSTAB_PATH: &str = "/etc/fstab";

/// Updates "/etc/fstab" to auto remount in case of instance reboot.
/// The output is the contents of the "fstab" file.
///
/// e.g.,
/// sudo echo '/dev/nvme1n1       /data   ext4    defaults,nofail 0       2' >> /etc/fstab
/// sudo mount --all
///
/// ref. See https://github.com/cholcombe973/block-utils/blob/master/src/lib.rs for other commands.
/// ref. https://stackoverflow.com/questions/45167717/mounting-a-nvme-disk-on-aws-ec2
pub fn update_fstab(
    filesystem_name: &str,
    device_name: &str,
    dir_name: &str,
) -> io::Result<(String, String)> {
    let device_path = if device_name.starts_with("/dev/") {
        device_name.to_string()
    } else {
        format!("/dev/{}", device_name)
    };

    log::info!(
        "updating the fstab file on the device path '{}'",
        device_path
    );

    let line = format!(
        "{}       {}   {}    defaults,nofail 0       2",
        device_path, dir_name, filesystem_name
    );
    let mut contents = fs::read_to_string(FSTAB_PATH)?;
    if contents.contains(&line) {
        log::warn!("fstab already contains '{}', skipping updating fstab", line);
        return Ok((contents, String::new()));
    }
    contents.push('\n');
    contents.push_str(&line);

    let tmp_path = random_manager::tmp_path(10, None)?;
    let mut f = File::create(&tmp_path)?;
    f.write_all(contents.as_bytes())?;

    let cmd = format!("sudo cp {} {}", tmp_path, FSTAB_PATH);
    command_manager::run(&cmd)?;
    command_manager::run("sudo mount --all")?;

    Ok((contents, String::new()))
}

File: linux/go/disk/disk.go
package disk

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/gyuho/infra/go/logutil"
	"github.com/gyuho/infra/go/randutil"

	"k8s.io/utils/exec"
)

// Makes a new file system on the specified device.
//
// e.g.,
// sudo mkfs -t ext4 /dev/nvme1n1
//
// Usually, "nvme0n1" is the boot volume.
// "nvme1n1" is the additional mounted volume.
//
// See https://github.com/cholcombe973/block-utils/blob/master/src/lib.rs for other commands.
// ref. https://stackoverflow.com/questions/45167717/mounting-a-nvme-disk-on-aws-ec2
func Mkfs(ctx context.Context, fsName string, deviceName string) ([]byte, error) {
	cmdPath, err := exec.New().LookPath("mkfs")
	if err != nil {
		return nil, fmt.Errorf("mkfs not found (%w)", err)
	}

	devicePath := deviceName
	if !strings.HasPrefix(deviceName, "/dev/") {
		devicePath = "/dev/" + deviceName
	}

	args := []string{cmdPath, "-t", fsName, devicePath}
	logutil.S().Infow("making a file system",
		"fsName", fsName,
		"devicePath", devicePath,
		"command", strings.Join(args, " "),
	)

	out, err := exec.New().CommandContext(ctx, args[0], args[1:]...).CombinedOutput()
	if err != nil {
		// e.g., mke2fs 1.45.5 (07-Jan-2020) /dev/nvme1n1 is mounted; will not make a filesystem here!
		// TODO: handle "mount: /data: wrong fs type, bad option, bad superblock on /dev/nvme1n1, missing codepage or helper program, or other error."
		if strings.Contains(string(out), devicePath+" is mounted") {
			logutil.S().Warnw("device is already mounted", "devicePath", devicePath, "error", err)
			return out, nil
		}
		return nil, err
	}
	return out, nil
}

// Mounts a file system to the specified directory.
//
// e.g.,
// sudo mount /dev/nvme1n1 /data -t ext4
//
// See https://github.com/cholcombe973/block-utils/blob/master/src/lib.rs for other commands.
// ref. https://stackoverflow.com/questions/45167717/mounting-a-nvme-disk-on-aws-ec2
func Mount(ctx context.Context, fsName string, deviceName string, dirPath string) ([]byte, error) {
	cmdPath, err := exec.New().LookPath("mount")
	if err != nil {
		return nil, fmt.Errorf("mount not found (%w)", err)
	}

	devicePath := deviceName
	if !strings.HasPrefix(deviceName, "/dev/") {
		devicePath = "/dev/" + deviceName
	}

	args := []string{cmdPath, devicePath, dirPath, fsName}
	logutil.S().Infow("mounting the file system",
		"fsName", fsName,
		"devicePath", devicePath,
		"dirPath", dirPath,
		"command", strings.Join(args, " "),
	)

	out, err := exec.New().CommandContext(ctx, args[0], args[1:]...).CombinedOutput()
	if err != nil {
		// e.g., mount: /data: /dev/nvme1n1 already mounted on /data
		if strings.Contains(string(out), devicePath+" already mounted") {
			logutil.S().Warnw("device is already mounted", "devicePath", devicePath, "error", err)
			return out, nil
		}
		return nil, err
	}
	return out, nil
}

const FSTAB_PATH = "/etc/fstab"

// Updates "/etc/fstab" to auto remount in case of instance reboot.
func UpdateFstab(ctx context.Context, fsName string, deviceName string, dirPath string) ([]byte, error) {
	devicePath := deviceName
	if !strings.HasPrefix(deviceName, "/dev/") {
		devicePath = "/dev/" + deviceName
	}

	line := fmt.Sprintf(`%s       %s   %s    defaults,nofail 0       2`, devicePath, dirPath, fsName)
	b, err := os.ReadFile(FSTAB_PATH)
	if err != nil {
		return nil, err
	}
	if strings.Contains(string(b), line) {
		logutil.S().Warnw("fstab already contains the line", "line", line)
		return b, nil
	}
	b = append(b, []byte("\n")...)
	b = append(b, []byte(line)...)

	tmpPath := filepath.Join(os.TempDir(), "fstab."+randutil.String(10)+".tmp")
	if err := os.WriteFile(tmpPath, b, 0644); err != nil {
		return nil, err
	}

	if _, err := exec.New().CommandContext(ctx, "cp", tmpPath, FSTAB_PATH).CombinedOutput(); err != nil {
		return nil, err
	}
	return b, nil
}
